// Copyright 2015, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

syntax = "proto3";

package csce438;

import "google/protobuf/timestamp.proto";

// The messenger service definition.
service SNSService{

  rpc Login (Request) returns (Reply) {}
  rpc List (Request) returns (Reply) {}
  rpc Follow (Request) returns (Reply) {}
  rpc ConnectClient (Request) returns (Reply) {}
  rpc Ping (Heartbeat) returns (Reply) {}
  rpc RegisterServer (ServerInfo) returns (ClusterInfo) {}
  rpc RegisterSync (SyncInfo) returns (ClusterInfo) {}
  rpc AddSync (SyncInfo) returns (ClusterInfo) {}
  rpc SyncFile (FileInfo) returns (Reply) {}
  rpc ConnectSlave (ServerInfo) returns (ClusterInfo) {}
  // Bidirectional streaming RPC
  rpc Timeline (stream Message) returns (stream Message) {}
  rpc UpdateTimeline (Request) returns (Reply) {} 
}

message Request {
  string id = 1;
  repeated string arguments = 2;
}

message Reply {
  string msg = 1;
  repeated string all_users = 2;
  repeated string followers = 3;
}

message Heartbeat {
  string id = 1;
  string type = 2;
  //google.protobuf.Timestamp timestamp = 3;
}

message ServerInfo {
  string id = 1;
  string type = 2;
  string port = 3;
}

message SyncInfo {
  string id = 1;
  string port = 2;
}

message ClusterInfo {
  string id = 1;
  string port = 2;
  repeated string syncs = 3;
}

message FileInfo {
  string id = 1;
  string action = 2;
  repeated string lines = 3;
}

message Message {
  //Username who sent the message
  string id = 1;
  //Message from the user
  string msg = 2;
  //Time the message was sent
  google.protobuf.Timestamp timestamp = 3;
}
